cpp/include/raft/spatial/knn/detail/ann_ivf_flat_kernel.cuh

```diff
@@ -0,0 +1,1227 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+// #define USE_FAISS
+
+#include "../ann_common.h"
+#include "ann_utils.cuh"
+#include "topk/warpsort_topk.cuh"
+
+#include <raft/common/device_loads_stores.cuh>
+#include <raft/core/logger.hpp>
+#include <raft/cuda_utils.cuh>
+#include <raft/cudart_utils.h>
+#include <raft/distance/distance.cuh>
+#include <raft/pow2_utils.cuh>
+
+#ifdef USE_FAISS
+#include <faiss/gpu/utils/Comparators.cuh>
+#include <faiss/gpu/utils/Select.cuh>
+#endif
+
+#include <rmm/cuda_stream_view.hpp>
+
+#include <optional>
+#include <raft/core/mdarray.hpp>
+
+namespace raft::spatial::knn::detail {
+
+constexpr int kThreadsPerBlock = 128;
+
+namespace md = std::experimental;
+
+template <typename T>
+struct ivf_flat_index {
+  using row_major = md::layout_right;
+  using extent_1d = md::extents<dynamic_extent>;
+  using extent_2d = md::extents<dynamic_extent, dynamic_extent>;
+
+  /**
+   * Vectorized load/store size in elements, determines the size of interleaved data chunks.
+   */
+  uint32_t veclen;
+
+  /**
+   * Inverted list data [size, dim].
+   *
+   * The data consists of the dataset rows, grouped by their labels (into clusters/lists).
+   * Within each list (cluster), the data is grouped into blocks of `WarpSize` interleaved
+   * vectors. Note, the total index length is slightly larger than the source dataset length,
+   * because each cluster is padded by `WarpSize` elements.
+   *
+   * Interleaving pattern:
+   * within groups of `WarpSize` rows, the data is interleaved with the block size equal to
+   * `veclen * sizeof(T)`. That is, a chunk of `veclen` consecutive components of one row is
+   * followed by a chunk of the same size of the next row, and so on.
+   *
+   * __Example__: veclen = 2, dim = 6, WarpSize = 32, list_size = 31
+   * `
+   *   x[ 0, 0], x[ 0, 1], x[ 1, 0], x[ 1, 1], ... x[14, 0], x[14, 1], x[15, 0], x[15, 1],
+   *   x[16, 0], x[16, 1], x[17, 0], x[17, 1], ... x[30, 0], x[30, 1],    -    ,    -    ,
+   *   x[ 0, 2], x[ 0, 3], x[ 1, 2], x[ 1, 3], ... x[14, 2], x[14, 3], x[15, 2], x[15, 3],
+   *   x[16, 2], x[16, 3], x[17, 2], x[17, 3], ... x[30, 2], x[30, 3],    -    ,    -    ,
+   *   x[ 0, 4], x[ 0, 5], x[ 1, 4], x[ 1, 5], ... x[14, 4], x[14, 5], x[15, 4], x[15, 5],
+   *   x[16, 4], x[16, 5], x[17, 4], x[17, 5], ... x[30, 4], x[30, 5],    -    ,    -    ,
+   * `
+   */
+  device_mdarray<T, extent_2d, row_major> data;
```

Comment:
Hmm, I've been thinking to follow our cuml dev guide here, which says to try to keep the structures POD/standard layout (modulo mdarray members). I believe it makes sense in this case for a few reasons: (1) to make it easier for third-parties to write their own serialization/deserialization and (2) to simplify cython wrappers if we choose to write them (like e.g. the linear_svm struct I added in cuml); (3) also it needs less code to maintain :).
At the same time, (4) I keep this structure `const` in the code, so there is no need to do any extra protection using private accessors. On the other hand though, there anyway shouldn't be a need to access the mdarray members except for the two methods we already have plus serialization.

That being said, If you still prefer to make it classy, I'm fine with that; please confirm.

Reply:
I wasn't thinking of this as a POD struct from the start, mostly because the bulk of its state is rather complex objects, and data pointers to device memory at that (and `mdarray` itself is not a POD, right?). I'm not completely against sticking w/ using the `check_consistency` method, but my proposal would 1) guarantee the consistency from the time of construction (like `mdarray` does), and 2) make the expectations more immediately obvious to future users of this object.

In terms of serialization, my proposal is more inline w/ the design of the corresponding FAISS indices and serialization frameworks (such as protobuf, as an example) generally provide mechanisms to adapt complex objects to make them serializable. I think we might need to do this anyways, at least for the `mdarray` type?

I think this is a fairly opaque object, since it's constructed by a RAFT function and intended to be used within a RAFT function. I guess the question then becomes, how often do we intend for users outside of RAFT to directly interact w/ its members?

Reply:
Following our other discussion below, I've refactored `build` and `search` into public functions and eliminated the handle type completely. The relevant types are now in [ivf_flat_types.hpp](https://github.com/rapidsai/raft/blob/df17b5b631ddec4e277c2af4386b9aa9f073bdd8/cpp/include/raft/spatial/knn/ivf_flat_types.hpp) and the functions are in [ivf_flat.cuh](https://github.com/rapidsai/raft/blob/df17b5b631ddec4e277c2af4386b9aa9f073bdd8/cpp/include/raft/spatial/knn/ivf_flat.cuh).

Back to our topic. In the current state, the users are not expected to construct the `index<T>` manually anyway, they should call the factory function `build<T>`, which ensures consistency. All public functionality is exposed in `ivf_flat.cuh` anyway, so I don't think there could be any difficulties with the expectations on the use of the type. Later we could add similar functions to serialize/deserialize etc to `ivf_flat.cuh`. At the same time, I can imagine a certain user wants to do access the raw index data for whatever reason; maybe they need a custom serialization (cannot use ours) or do some modifications to the computed clusters. All of the members (mdarrays) are constructible from outside raft, so this is possible now. If we choose to switch to the opaque object, all of these would require us to write more code.

Also, personally I'd prefer to have a more homogenous api: either make everything via object methods or keep everything as public functions. I have a feeling that adding constructors, serialization, and an array of accessor methods would go against that in the presence of our `build` and `search` public functions.

Reply:
> Also, personally I'd prefer to have a more homogenous api: either make everything via object methods or keep everything as public functions. 

RAFT's core public APIs are flat public functions. That said, there are times when we use more complex objects which provide proper constructors and methods to maintain consistency of their own internal state (thinking `handle`, `mdspan`, `mdarray` here). In general, we have already been following the homogenous design that you mention but the `cuivfhandle` breaks this design by exposing public methods on the handle itself. That design isn't really necessary anyways because it's just being invoked from flattened public functions anyways. I don't think we need the extra layer there.

Reply:
> Following our other discussion below, I've refactored build and search into public functions and eliminated the handle type completely. 

This looks much better.